source/modulo_components/include/modulo_components/ComponentInterface.h

```diff
@@ -19,19 +20,58 @@
 namespace modulo_components {
 
 template<class NodeT, typename PubT>
-class ComponentInterface : public state_representation::ParameterMap, public NodeT {
+class ComponentInterface : NodeT {
+  friend class TestComponentInterface;
+  FRIEND_TEST(TestComponentInterface, test_add_bool_predicate);
+  FRIEND_TEST(TestComponentInterface, test_add_function_predicate);
+  FRIEND_TEST(TestComponentInterface, test_get_predicate_value);
+  FRIEND_TEST(TestComponentInterface, test_set_predicate_value);
```

Comment:
if the fixture class is declared as a friend, I would prefer you define the appropriate visibility in that class rather than declaring each test case as a friend. This otherwise won't scale well. 
```suggestion
  friend class ComponentInterfaceTest;
```

Reply:
I followed the example from the link you sent. I will try to check if that is sufficient and can work like this. Indeed that would be better.

Reply:
I made the test and as I thought this does not work. If you want to call a protected function in a test function, this test function needs to be explicitly added as `FRIEND_TEST`

Reply:
It would definitely work fine as long as the `ComponentInterfaceTest` fixture class provides the right accessors. Better for the test code to scale with each test requirement than the class header. I tested it locally to make sure that something like the following works if you only declare `friend class ComponentInterfaceTest;` and nothing else.

It also raises the question if these functions, which are designed to be a sort of user-interface for the developer, really need to be `protected` in the first place. It's rare that any user will want access to an instance of the component at run-time, and certainly is not possible from within our application at run-time. 

Anyway TL;DR I think expanding a wrapper class in the test suite is preferable to expanding friends for test cases in the public header.

```c++
#include ...

namespace modulo_components {
class ComponentInterfaceTest : public ::testing::Test {
protected:
  static void SetUpTestSuite() {
    rclcpp::init(0, nullptr);
  }

  void SetUp() override {
    component_ = std::make_shared<ComponentInterface<rclcpp::Node, rclcpp::Publisher<modulo_new_core::EncodedState>>>(rclcpp::NodeOptions());
  }

  void TearDown() override {
    component_->predicates_.clear();
  }

  std::map<std::string, utilities::PredicateVariant> get_predicate_map() {
    return component_->predicates_;
  }

  void add_predicate(const std::string& name, bool value) {
    component_->add_predicate(name, value);
  }

  void add_predicate(const std::string& predicate_name, const std::function<bool(void)>& predicate_function) {
    component_->add_predicate(predicate_name, predicate_function);
  }

  [[nodiscard]] bool get_predicate(const std::string& predicate_name) const {
    return component_->get_predicate(predicate_name);
  }

  /**
   * @brief Set the value of the predicate given as parameter
   * @param predicate_name the name of the predicate to retrieve from the
   * map of predicates
   * @param predicate_value the new value of the predicate
   */
  void set_predicate(const std::string& predicate_name, bool predicate_value) {
    component_->set_predicate(predicate_name, predicate_value);
  }

  std::shared_ptr<ComponentInterface<rclcpp::Node, rclcpp::Publisher<modulo_new_core::EncodedState>>> component_;
};

TEST_F(ComponentInterfaceTest, test_add_bool_predicate) {
  add_predicate("foo", true);
  auto predicates = get_predicate_map();
  auto predicate_iterator = predicates.find("foo");
  EXPECT_TRUE(predicate_iterator != predicates.end());
  auto value = std::get<bool>(predicate_iterator->second);
  EXPECT_TRUE(value);
}

TEST_F(ComponentInterfaceTest, test_add_function_predicate) {
  add_predicate("bar", [&]() { return false; });
  auto predicates = get_predicate_map();
  auto predicate_iterator = predicates.find("bar");
  EXPECT_TRUE(predicate_iterator != predicates.end());
  auto value_callback = std::get<std::function<bool(void)>>(predicate_iterator->second);
  EXPECT_FALSE((value_callback)());
}

TEST_F(ComponentInterfaceTest, test_get_predicate_value) {
  add_predicate("foo", true);
  EXPECT_TRUE(get_predicate("foo"));
  add_predicate("bar", [&]() { return false; });
  EXPECT_FALSE(get_predicate("bar"));
  // TODO suppress the no discard warning
  EXPECT_THROW(get_predicate("test"), exceptions::PredicateNotFoundException);
}

TEST_F(ComponentInterfaceTest, test_set_predicate_value) {
  add_predicate("foo", true);
  set_predicate("foo", false);
  EXPECT_FALSE(get_predicate("foo"));
  EXPECT_THROW(set_predicate("bar", true), exceptions::PredicateNotFoundException);
  add_predicate("bar", [&]() { return false; });
  set_predicate("bar", true);
  EXPECT_TRUE(get_predicate("bar"));
}
}
```

Reply:
Right that is actually much better. I was a bit slow on the understanding there.